gpbft/gpbft_test.go

```diff
@@ -0,0 +1,457 @@
+package gpbft_test
+
+import (
+	"testing"
+
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/filecoin-project/go-f3/gpbft/emulator"
+)
+
+var (
+	tipset0 = gpbft.TipSet{Epoch: 0, Key: []byte("bigbang")}
+	tipSet1 = gpbft.TipSet{Epoch: 1, Key: []byte("fish")}
+	tipSet2 = gpbft.TipSet{Epoch: 2, Key: []byte("lobster")}
+)
+
+func TestGPBFT_DecidesProposalOnStrongEvidenceOfPrepare(t *testing.T) {
+	require := emulator.NewAssertions(t)
+
+	instance, err := emulator.NewInstance(
+		0,
+		gpbft.PowerEntries{
+			gpbft.PowerEntry{
+				ID:    0,
+				Power: gpbft.NewStoragePower(1),
+			},
+			gpbft.PowerEntry{
+				ID:    1,
+				Power: gpbft.NewStoragePower(1),
+			},
+		},
+		tipset0, tipSet1, tipSet2,
+	)
+	require.NoError(err)
+
+	em, err := emulator.New()
+	require.NoError(err)
+
+	// Start the instance and trigger the first alarm because the GPBFT
+	// implementation in this repo implicitly uses the alarm to actually start the
+	// instance.
+	require.NoError(em.Start(instance))
+	require.AlarmReceived(em.ReceiveAlarm())
+
+	// Expect QUALITY step from the emulator.
+	require.Quality(require.MessageReceived(em.ReceiveBroadcast()), instance)
```

Comment:
Interesting, this is somewhat inverted from what I expected. I expected:
1. Assert that QUALITY was sent
2. Deliver a QUALITY message

This is kind of inferring that a message was sent by delivering an unknown message that the participant sent earlier and then checking that it exists and was QUALITY. It works, but this scheme will always deliver participants own message first, and seems awkward to re-order.

Reply:
>It works, but this scheme will always deliver participants own message first, and seems awkward to re-order.

Happy to add inspection of what the honest participant requested the broadcast of before delivering it. I don't think it makes much difference because we assert that what we delivered was what we expected to deliver immediately after it is delivered.